File: 200-number-of-islands/200-number-of-islands.c
void dfs(char **grid, int gridSize, int gridColSize, int i, int j) {
    if (i == -1 || j == -1) {
        return;
    }
    
    if (i > gridSize - 1 || j > gridColSize - 1) {
        return;
    }
    
    if (grid[i][j] == '0') {
        return;
    }

    grid[i][j] = '0';
    dfs(grid, gridSize, gridColSize, i-1, j);
    dfs(grid, gridSize, gridColSize, i, j-1);
    dfs(grid, gridSize, gridColSize, i+1, j);
    dfs(grid, gridSize, gridColSize, i, j+1);
}

int numIslands(char** grid, int gridSize, int* gridColSize){
    int i,j;
    int result = 0;
    
    for (i = 0; i < gridSize; i++) {
        for (j = 0; j < gridColSize[i]; j++) {
            if (grid[i][j] == '1') {
                dfs(grid, gridSize, gridColSize[i], i, j);
                result++;
            }            
        }
    }
    
    return result;
}

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:        
        def dfs(r: int, c: int):
            if r < 0 or c < 0 or r >= len(grid) or c >= len(grid[0]):
                return
            
            if grid[r][c] == "0":
                return
            
            grid[r][c] = "0"
            
            dfs(r+1, c)
            dfs(r, c+1)
            dfs(r-1, c)
            dfs(r, c-1)

        result = 0
        
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == "1":
                    result+=1
                    dfs(r,c)

        return result